pkg/server/registration/manager.go

```diff
@@ -0,0 +1,91 @@
+package registration
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/pkg/common/telemetry"
+	telemetry_server "github.com/spiffe/spire/pkg/common/telemetry/server"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/proto/spire/server/datastore"
+)
+
+const (
+	_pruningCandence = 2 * time.Hour
+)
+
+// ManagerConfig is the config for the registration manager
+type ManagerConfig struct {
+	DataStore datastore.DataStore
+
+	Log     logrus.FieldLogger
+	Metrics telemetry.Metrics
+
+	Clock clock.Clock
+}
+
+// Manager is the manager of registrations
+type Manager struct {
+	c       ManagerConfig
+	log     logrus.FieldLogger
+	metrics telemetry.Metrics
+}
+
+// NewManager creates a new registration manager
+func NewManager(c ManagerConfig) *Manager {
+	if c.Clock == nil {
+		c.Clock = clock.New()
+	}
+
+	return &Manager{
+		c:       c,
+		log:     c.Log.WithField(telemetry.RetryInterval, _pruningCandence),
+		metrics: c.Metrics,
+	}
+}
+
+// Run runs the registration manager
+func (m *Manager) Run(ctx context.Context) error {
+	err := util.RunTasks(ctx,
+		func(ctx context.Context) error {
+			return m.pruneEvery(ctx)
+		},
+	)
+	if err == context.Canceled {
+		err = nil
+	}
+	return err
+}
+
+func (m *Manager) pruneEvery(ctx context.Context) error {
+	ticker := m.c.Clock.Ticker(_pruningCandence)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			m.prune(ctx)
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (m *Manager) prune(ctx context.Context) (err error) {
+	counter := telemetry_server.StartRegistrationManagerPruneEntryCall(m.c.Metrics)
+	defer counter.Done(&err)
+
+	// drop response output
+	_, err = m.c.DataStore.PruneRegistrationEntries(ctx, &datastore.PruneRegistrationEntriesRequest{
+		ExpiresBefore: m.c.Clock.Now().Unix(),
+	})
+
+	if err != nil {
+		return fmt.Errorf("unable to prune registration entries: %v", err)
```

Comment:
This error is not actually printed anywhere. Was the intention to log it?

Related note: this condition can be hit if we are shutting down, or the context is otherwise cancelled, while in the middle of a prune operation. We should probably check that the context is still good before surfacing any error messages?

Reply:
it seems like `server.go` already handles this:
```
	err = util.RunTasks(ctx,
		caManager.Run,
		svidRotator.Run,
		endpointsServer.ListenAndServe,
		metrics.ListenAndServe,
		bundleManager.Run,
		registrationManager.Run,
		healthChecks.ListenAndServe,
	)
	if err == context.Canceled {
		err = nil
	}
```

if any error occurs in the processes, the server stops. If the error was due to cancelled context, no error returned. Else return the error(s)

Reply:
nvm I misunderstood the block above :)